### PR TITLE
Change `ResultShunt` to be generic over `Try`

### PR DESCRIPTION
**Pull request by scottmcm**
Just a refactor (and rename) for now, so it's not `Result`-specific.

This could be used for a future `Iterator::try_collect`, or similar, but anything like that is left for a future PR.

### CODE REVIEW

File: library/core/src/iter/adapters/mod.rs [hunk 1]
```diff
@@ -128,59 +128,62 @@ pub unsafe trait SourceIter {
...
 
-    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    fn try_fold<B, F, T>(&mut self, init: B, mut f: F) -> T
     where
-        F: FnMut(B, Self::Item) -> R,
-        R: Try<Output = B>,
+        F: FnMut(B, Self::Item) -> T,
+        T: Try<Output = B>,
     {
-        let error = &mut *self.error;
```
Comment (scottmcm) [inline library/core/src/iter/adapters/mod.rs:178]: This change needs the 2021 closure rules, which is why the change to `core-no-fp-fmt-parse` is needed.

File: library/core/src/iter/traits/accum.rs [hunk 2]
```diff
@@ -210,7 +210,7 @@ where
     where
         I: Iterator<Item = Option<U>>,
     {
-        iter.map(|x| x.ok_or(())).sum::<Result<_, _>>().ok()
+        iter::try_process(iter, |i| i.sum())
```
Comment (yaahc) [inline library/core/src/iter/traits/accum.rs:213]: you love to see the simplifications

File: library/core/src/iter/adapters/mod.rs (hunk 1 above)
Comment (yaahc) [reply library/core/src/iter/adapters/mod.rs:178]: Could this be avoided by including a similar line for the `self.residual` capture? I'm not sure it's worth trying to avoid but I want to make sure that I understand completely.

File: library/core/src/iter/adapters/mod.rs [hunk 3]
```diff
@@ -128,59 +128,62 @@ pub unsafe trait SourceIter {
...
 where
-    I: Iterator<Item = Result<T, E>>,
+    I: Iterator<Item: Try<Residual = R>>,
 {
-    type Item = T;
+    type Item = <I::Item as Try>::Output;
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.find(|_| true)
+        self.try_for_each(ControlFlow::Break).break_value()
```
Comment (yaahc) [inline library/core/src/iter/adapters/mod.rs:165]: Not important for this PR, but is there a reason we went with `break_value` for this method name instead of just `break`? The latter seems more consistent with the equivalent APIs on Result.

**Review by yaahc — COMMENTED** [review]
very fun. I'm pretty much ready to approve, just a couple minor comments.

File: library/core/src/iter/adapters/mod.rs (hunk 1 above)
Comment (scottmcm) [reply library/core/src/iter/adapters/mod.rs:178]: Yes, it absolutely could do the capture splitting manually, and thus continue to work on 2018.

But the normal build of core is already on 2021:

https://github.com/rust-lang/rust/blob/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/core/Cargo.toml#L9

so it seems weird to me to avoid using 2021 features, and thus it's better to make the run-make test consistent.

File: library/core/src/iter/adapters/mod.rs (hunk 3 above)
Comment (scottmcm) [reply library/core/src/iter/adapters/mod.rs:165]: Because `break` is a keyword:
```
1 | fn break() {}
  |    ^^^^^ expected identifier, found keyword
```

I agree, though, in general.  "What methods make sense" is still open in #75744

(The current name was inspired by [loop break value](https://github.com/rust-lang/rfcs/blob/master/text/1624-loop-break-value.md)).

Comment (yaahc) [reply library/core/src/iter/adapters/mod.rs:165]: gah, of course.

**Comment by matthiaskrgr** [comment]
Failed in a rollup: https://github.com/rust-lang/rust/pull/93702#issuecomment-1030874871
@bors r-

**Comment by scottmcm** [comment]
Fixes for the 2021 stuff are now in #93719 -- I'll rebase this onto that later.

**Comment by scottmcm** [comment]
@bors r=yaahc

No new code changes; just rebased onto master now that my other PR landed.
